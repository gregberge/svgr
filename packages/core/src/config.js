import { cosmiconfig, cosmiconfigSync } from 'cosmiconfig'

export const DEFAULT_CONFIG = {
  dimensions: true,
  expandProps: 'end',
  icon: false,
  native: false,
  typescript: false,
  prettier: true,
  prettierConfig: null,
  memo: false,
  ref: false,
  replaceAttrValues: null,
  svgProps: null,
  svgo: true,
  svgoConfig: { plugins: [{ name: 'convertStyleToAttrs', active: true }]},
  template: null,
  titleProp: false,
  runtimeConfig: true,
  plugins: null,
  namedExport: 'ReactComponent',
}

const explorer = cosmiconfig('svgr', {
  sync: true,
  cache: true,
  rcExtensions: true,
})

const explorerSync = cosmiconfigSync('svgr', {
  sync: true,
  cache: true,
  rcExtensions: true,
})

export async function resolveConfig(searchFrom, configFile) {
  if (configFile == null) {
    const result = await explorer.search(searchFrom)
    return result ? result.config : null
  }
  const result = await explorer.load(configFile)
  return result ? result.config : null
}

resolveConfig.sync = (searchFrom, configFile) => {
  if (configFile == null) {
    const result = explorerSync.search(searchFrom)
    return result ? result.config : null
  }
  const result = explorerSync.load(configFile)
  return result ? result.config : null
}

export async function resolveConfigFile(filePath) {
  const result = await explorer.search(filePath)
  return result ? result.filepath : null
}

resolveConfigFile.sync = (filePath) => {
  const result = explorerSync.search(filePath)
  return result ? result.filepath : null
}

export async function loadConfig({ configFile, ...baseConfig }, state = {}) {
  const rcConfig =
    state.filePath && baseConfig.runtimeConfig !== false
      ? await resolveConfig(state.filePath, configFile)
      : {}
  return { ...DEFAULT_CONFIG, ...rcConfig, ...baseConfig }
}

loadConfig.sync = ({ configFile, ...baseConfig }, state = {}) => {
  const rcConfig =
    state.filePath && baseConfig.runtimeConfig !== false
      ? resolveConfig.sync(state.filePath, configFile)
      : {}
  return { ...DEFAULT_CONFIG, ...rcConfig, ...baseConfig }
}
