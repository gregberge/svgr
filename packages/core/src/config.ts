import { cosmiconfig, cosmiconfigSync } from 'cosmiconfig'
import type { Options as PrettierOptions } from 'prettier'
import type { Config as SvgoConfig } from 'svgo'
import type { Options as TransformOptions } from '@svgr/babel-preset'
import type { TransformOptions as BabelTransformOptions } from '@babel/core'
import type { ConfigPlugin } from './plugins'
import type { State } from './state'

export interface Config {
  ref?: boolean
  titleProp?: boolean
  descProp?: boolean
  expandProps?: boolean | 'start' | 'end'
  dimensions?: boolean
  icon?: boolean | string | number
  native?: boolean
  svgProps?: {
    [key: string]: string
  }
  replaceAttrValues?: {
    [key: string]: string
  }
  runtimeConfig?: boolean
  typescript?: boolean
  prettier?: boolean
  prettierConfig?: PrettierOptions
  svgo?: boolean
  svgoConfig?: SvgoConfig
  configFile?: string
  template?: TransformOptions['template']
  memo?: boolean
  exportType?: 'named' | 'default'
  namedExport?: string
  jsxRuntime?: 'classic' | 'classic-preact' | 'automatic'
  jsxRuntimeImport?: {
    source: string
    namespace?: string
    specifiers?: string[]
    defaultSpecifier?: string
  }

  // CLI only
  index?: boolean
  plugins?: ConfigPlugin[]

  // JSX
  jsx?: {
    babelConfig?: BabelTransformOptions
  }
}

export const DEFAULT_CONFIG: Config = {
  dimensions: true,
  expandProps: 'end',
  icon: false,
  native: false,
  typescript: false,
  prettier: true,
  prettierConfig: undefined,
  memo: false,
  ref: false,
  replaceAttrValues: undefined,
  svgProps: undefined,
  svgo: true,
  svgoConfig: undefined,
  template: undefined,
  index: false,
  titleProp: false,
  descProp: false,
  runtimeConfig: true,
  namedExport: 'ReactComponent',
  exportType: 'default',
}

const explorer = cosmiconfig('svgr')
const explorerSync = cosmiconfigSync('svgr')

export const resolveConfig = async (
  searchFrom?: string,
  configFile?: string,
): Promise<Config | null> => {
  if (configFile == null) {
    const result = await explorer.search(searchFrom)
    return result ? result.config : null
  }
  const result = await explorer.load(configFile)
  return result ? result.config : null
}

resolveConfig.sync = (
  searchFrom?: string,
  configFile?: string,
): Config | null => {
  if (configFile == null) {
    const result = explorerSync.search(searchFrom)
    return result ? result.config : null
  }
  const result = explorerSync.load(configFile)
  return result ? result.config : null
}

export const resolveConfigFile = async (
  filePath: string,
): Promise<string | null> => {
  const result = await explorer.search(filePath)
  return result ? result.filepath : null
}

resolveConfigFile.sync = (filePath: string): string | null => {
  const result = explorerSync.search(filePath)
  return result ? result.filepath : null
}

export const loadConfig = async (
  { configFile, ...baseConfig }: Config,
  state: Pick<State, 'filePath'> = {},
): Promise<Config> => {
  const rcConfig =
    state.filePath && baseConfig.runtimeConfig !== false
      ? await resolveConfig(state.filePath, configFile)
      : {}
  return { ...DEFAULT_CONFIG, ...baseConfig, ...rcConfig }
}

loadConfig.sync = (
  { configFile, ...baseConfig }: Config,
  state: Pick<State, 'filePath'> = {},
): Config => {
  const rcConfig =
    state.filePath && baseConfig.runtimeConfig !== false
      ? resolveConfig.sync(state.filePath, configFile)
      : {}
  return { ...DEFAULT_CONFIG, ...baseConfig, ...rcConfig }
}
