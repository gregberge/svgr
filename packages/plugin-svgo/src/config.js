import { extendDefaultPlugins } from 'svgo';
import deepmerge from 'deepmerge'

export function getFilePath(state) {
  return state.filePath || process.cwd()
}

export function getBaseSvgoConfig(config) {
  const baseSvgoConfig = {
    plugins: [{ name: 'prefixIds', active: true }],
  }
  if (config.icon || config.dimensions === false) {
    baseSvgoConfig.plugins.push({ name: 'removeViewBox', active: false })
  }

  return {
      plugins: config.full ? baseSvgoConfig.plugins : extendDefaultPlugins(baseSvgoConfig.plugins)
  }
}

export function getPlugins(config) {
  if (!config || !config.plugins) {
    return []
  }
  if (!Array.isArray(config.plugins)) {
    throw Error('`svgoConfig.plugins` must be an array')
  }
  return config.plugins
}

function extractPlugins(config) {
  if (!config) return []
  if (!config.plugins) return []
  if (!Array.isArray(config.plugins)) return [config.plugins]
  return config.plugins
}

function mergePlugins(configs) {
  return configs
      .flatMap(extractPlugins)
      .reduce((plugins, plugin) => {
        const index = plugins.findIndex(accPlugin => accPlugin.name === plugin.name);
        if (index > -1) {
          plugins[index] = deepmerge(plugins[index], plugin)
        } else {
          plugins = [...plugins, plugin];
        }

        return plugins;
      }, []);
}

export function mergeSvgoConfig(...configs) {
  const plugins = mergePlugins(configs)
  return { ...deepmerge.all(configs.filter(Boolean)), plugins }
}
