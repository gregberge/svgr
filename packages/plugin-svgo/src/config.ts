import { cosmiconfigSync } from 'cosmiconfig'
import type { Config, State } from '@svgr/core'

const explorer = cosmiconfigSync('svgo', {
  searchPlaces: [
    'package.json',
    '.svgorc',
    '.svgorc.js',
    '.svgorc.json',
    '.svgorc.yaml',
    '.svgorc.yml',
    'svgo.config.js',
    'svgo.config.cjs',
    '.svgo.yml',
  ],
  transform: (result) => result && result.config,
  cache: true,
})

const getSvgoConfigFromSvgrConfig = (config: Config): any => {
  const params = { overrides: {} as any }
  if (config.icon || config.dimensions === false) {
    params.overrides.removeViewBox = false
  }
  if (config.native) {
    params.overrides.inlineStyles = {
      onlyMatchedOnce: false,
    }
  }

  return {
    plugins: [
      {
        name: 'preset-default',
        params,
      },
      'prefixIds',
    ],
  }
}

export const getSvgoConfig = (config: Config, state: State): any => {
  const cwd = state.filePath || process.cwd()
  if (config.svgoConfig) return config.svgoConfig
  if (config.runtimeConfig) {
    const userConfig = explorer.search(cwd)
    if (userConfig) return userConfig
  }
  return getSvgoConfigFromSvgrConfig(config)
}
