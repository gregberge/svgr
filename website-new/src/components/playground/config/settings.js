const parseObject = (value) =>
  value.split(',').reduce((obj, assignment) => {
    const [left, right] = assignment.split('=')
    return {
      ...obj,
      [left]: right,
    }
  }, {})

const parseJson = (value) => value && JSON.parse(value)

const initialSvgoConfig = JSON.stringify(
  {
    plugins: [
      {
        name: 'preset-default',
        params: { overrides: { removeTitle: false } },
      },
    ],
  },
  null,
  2,
)

const initialPrettierConfig = JSON.stringify({ semi: false }, null, 2)

export const settings = [
  {
    label: 'Icon',
    name: 'icon',
    type: 'boolean',
    group: 'global',
    default: false,
  },
  {
    label: 'React Native',
    name: 'native',
    type: 'boolean',
    group: 'global',
    default: false,
  },
  {
    label: 'TypeScript',
    name: 'typescript',
    type: 'boolean',
    group: 'global',
    default: false,
  },
  {
    label: 'Ref',
    name: 'ref',
    type: 'boolean',
    group: 'global',
    default: false,
  },
  {
    label: 'Memo',
    name: 'memo',
    type: 'boolean',
    group: 'global',
    default: false,
  },
  {
    label: 'Title prop',
    name: 'titleProp',
    type: 'boolean',
    group: 'global',
    default: false,
  },
  {
    label: 'Desc prop',
    name: 'descProp',
    type: 'boolean',
    group: 'global',
    default: false,
  },
  {
    label: 'Expand props',
    name: 'expandProps',
    type: 'enum',
    values: ['start', 'end', 'none'],
    group: 'global',
    default: 'end',
    transform: (value) => (value === 'none' ? false : value),
  },
  {
    label: 'Replace attributes value',
    name: 'replaceAttrValues',
    placeholder: '#063855=currentColor',
    type: 'string',
    group: 'global',
    transform: parseObject,
    default: '',
  },
  {
    label: 'SVG props',
    name: 'svgProps',
    placeholder: 'focusable=false',
    type: 'string',
    group: 'global',
    transform: parseObject,
    default: '',
  },
  {
    label: 'Enable',
    name: 'svgo',
    type: 'boolean',
    group: 'svgo',
    default: true,
  },
  {
    label: 'Config',
    name: 'svgoConfig',
    placeholder: initialSvgoConfig,
    type: 'string',
    group: 'svgo',
    default: initialSvgoConfig,
    transform: parseJson,
  },
  {
    label: 'Enable',
    name: 'prettier',
    type: 'boolean',
    group: 'prettier',
    default: true,
  },
  {
    label: 'Config',
    name: 'prettierConfig',
    placeholder: initialPrettierConfig,
    type: 'string',
    group: 'prettier',
    default: initialPrettierConfig,
    transform: parseJson,
  },
]

export const getInitialState = () =>
  settings.reduce(
    (state, setting) => ({
      ...state,
      [setting.name]: setting.default,
    }),
    {},
  )

export function stateToSettings(state) {
  return Object.entries(state).reduce((newState, [key, value]) => {
    const { transform } = settings.find(({ name }) => name === key)
    return { ...newState, [key]: transform ? transform(value) : value }
  }, {})
}
