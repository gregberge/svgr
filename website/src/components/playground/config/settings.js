const parseObject = value =>
  value.split(',').reduce((obj, assignment) => {
    const [left, right] = assignment.split('=')
    return {
      ...obj,
      [left]: right,
    }
  }, {})

const parseJson = value => value && JSON.parse(value)

export const settings = [
  {
    label: 'Icon',
    name: 'icon',
    type: 'boolean',
    group: 'global',
    default: false,
  },
  {
    label: 'React Native',
    name: 'native',
    type: 'boolean',
    group: 'global',
    default: false,
  },
  {
    label: 'Ref',
    name: 'ref',
    type: 'boolean',
    group: 'global',
    default: false,
  },
  {
    label: 'Title prop',
    name: 'titleProp',
    type: 'boolean',
    group: 'global',
    default: false,
  },
  {
    label: 'Expand props',
    name: 'expandProps',
    type: 'enum',
    values: ['start', 'end', 'none'],
    group: 'global',
    default: 'end',
    transform: value => (value === 'none' ? false : value),
  },
  {
    label: 'Replace attributes value',
    name: 'replaceAttrValues',
    placeholder: '#063855=currentColor',
    type: 'string',
    group: 'global',
    transform: parseObject,
    default: '',
  },
  {
    label: 'SVG props',
    name: 'svgProps',
    placeholder: 'focusable=false',
    type: 'string',
    group: 'global',
    transform: parseObject,
    default: '',
  },
  {
    label: 'Enable',
    name: 'svgo',
    type: 'boolean',
    group: 'svgo',
    default: true,
  },
  {
    label: 'Config',
    name: 'svgoConfig',
    placeholder: '{ "plugins": [{ "removeTitle": false }] }',
    type: 'string',
    group: 'svgo',
    default: '{ "plugins": [{ "removeTitle": false }] }',
    transform: parseJson,
  },
  {
    label: 'Enable',
    name: 'prettier',
    type: 'boolean',
    group: 'prettier',
    default: true,
  },
  {
    label: 'Config',
    name: 'prettierConfig',
    placeholder: '{ "semi": false }',
    type: 'string',
    group: 'prettier',
    default: '{ "semi": false }',
    transform: parseJson,
  },
]

export const getInitialState = () =>
  settings.reduce(
    (state, setting) => ({
      ...state,
      [setting.name]: setting.default,
    }),
    {},
  )

export const transformSettings = state =>
  Object.entries(state).reduce((newState, [key, value]) => {
    const { transform } = settings.find(({ name }) => name === key)
    return { ...newState, [key]: transform ? transform(value) : value }
  }, {})
