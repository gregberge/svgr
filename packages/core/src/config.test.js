import path from 'path'
import { resolveConfig, resolveConfigFile } from './config'

describe('config', () => {
  describe('#resolveConfig', () => {
    it('should return null if no config found', async () => {
      const config = await resolveConfig('/tmp')
      expect(config).toBe(null)
    })

    it('should return config if found', async () => {
      const config = await resolveConfig(
        path.join(__dirname, '__fixtures__/svgr'),
      )
      expect(config).toEqual({
        icon: true,
        noSemi: true,
        replaceAttrValues: [['#063855', 'currentColor']],
      })
    })
  })

  describe('#resolveConfigFile', () => {
    it('should return null if no config found', async () => {
      const config = await resolveConfigFile('/tmp')
      expect(config).toBe(null)
    })

    it('should return config path if found', async () => {
      const config = await resolveConfigFile(
        path.join(__dirname, '__fixtures__/svgr'),
      )
      expect(config).toMatch(/__fixtures__\/svgr\/\.svgrrc$/)
    })
  })
})
