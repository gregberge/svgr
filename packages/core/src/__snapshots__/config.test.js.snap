// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`svgo async #loadConfig [async] should load config using filePath 1`] = `
Object {
  "dimensions": true,
  "expandProps": "end",
  "icon": true,
  "memo": false,
  "namedExport": "ReactComponent",
  "native": false,
  "noSemi": true,
  "plugins": null,
  "prettier": true,
  "prettierConfig": null,
  "ref": false,
  "replaceAttrValues": Array [
    Array [
      "#063855",
      "currentColor",
    ],
  ],
  "runtimeConfig": true,
  "svgProps": null,
  "svgo": true,
  "svgoConfig": Object {
    "plugins": Array [
      Object {
        "active": true,
        "name": "convertStyleToAttrs",
      },
    ],
  },
  "template": null,
  "titleProp": false,
  "typescript": false,
}
`;

exports[`svgo async #loadConfig [async] should not load config with "runtimeConfig: false 1`] = `
Object {
  "dimensions": true,
  "expandProps": "end",
  "icon": true,
  "memo": false,
  "namedExport": "ReactComponent",
  "native": false,
  "noSemi": true,
  "plugins": null,
  "prettier": true,
  "prettierConfig": null,
  "ref": false,
  "replaceAttrValues": Array [
    Array [
      "#063855",
      "currentColor",
    ],
  ],
  "runtimeConfig": true,
  "svgProps": null,
  "svgo": true,
  "svgoConfig": Object {
    "plugins": Array [
      Object {
        "active": true,
        "name": "convertStyleToAttrs",
      },
    ],
  },
  "template": null,
  "titleProp": false,
  "typescript": false,
  "useRuntimeConfig": false,
}
`;

exports[`svgo async #loadConfig [async] should use default config without state.filePath 1`] = `
Object {
  "dimensions": false,
  "expandProps": "end",
  "icon": false,
  "memo": false,
  "namedExport": "ReactComponent",
  "native": false,
  "plugins": null,
  "prettier": true,
  "prettierConfig": null,
  "ref": false,
  "replaceAttrValues": null,
  "runtimeConfig": true,
  "svgProps": null,
  "svgo": true,
  "svgoConfig": Object {
    "plugins": Array [
      Object {
        "active": true,
        "name": "convertStyleToAttrs",
      },
    ],
  },
  "template": null,
  "titleProp": false,
  "typescript": false,
}
`;

exports[`svgo async #loadConfig [async] should work with custom config path 1`] = `
Object {
  "dimensions": true,
  "expandProps": "end",
  "icon": true,
  "memo": false,
  "namedExport": "ReactComponent",
  "native": false,
  "noSemi": true,
  "plugins": null,
  "prettier": true,
  "prettierConfig": null,
  "ref": false,
  "replaceAttrValues": Array [
    Array [
      "#063855",
      "currentColor",
    ],
  ],
  "runtimeConfig": true,
  "svgProps": null,
  "svgo": true,
  "svgoConfig": Object {
    "plugins": Array [
      Object {
        "active": true,
        "name": "convertStyleToAttrs",
      },
    ],
  },
  "template": null,
  "titleProp": false,
  "typescript": false,
}
`;

exports[`svgo sync #loadConfig [sync] should load config using filePath 1`] = `
Object {
  "dimensions": true,
  "expandProps": "end",
  "icon": true,
  "memo": false,
  "namedExport": "ReactComponent",
  "native": false,
  "noSemi": true,
  "plugins": null,
  "prettier": true,
  "prettierConfig": null,
  "ref": false,
  "replaceAttrValues": Array [
    Array [
      "#063855",
      "currentColor",
    ],
  ],
  "runtimeConfig": true,
  "svgProps": null,
  "svgo": true,
  "svgoConfig": Object {
    "plugins": Array [
      Object {
        "active": true,
        "name": "convertStyleToAttrs",
      },
    ],
  },
  "template": null,
  "titleProp": false,
  "typescript": false,
}
`;

exports[`svgo sync #loadConfig [sync] should not load config with "runtimeConfig: false 1`] = `
Object {
  "dimensions": true,
  "expandProps": "end",
  "icon": true,
  "memo": false,
  "namedExport": "ReactComponent",
  "native": false,
  "noSemi": true,
  "plugins": null,
  "prettier": true,
  "prettierConfig": null,
  "ref": false,
  "replaceAttrValues": Array [
    Array [
      "#063855",
      "currentColor",
    ],
  ],
  "runtimeConfig": true,
  "svgProps": null,
  "svgo": true,
  "svgoConfig": Object {
    "plugins": Array [
      Object {
        "active": true,
        "name": "convertStyleToAttrs",
      },
    ],
  },
  "template": null,
  "titleProp": false,
  "typescript": false,
  "useRuntimeConfig": false,
}
`;

exports[`svgo sync #loadConfig [sync] should use default config without state.filePath 1`] = `
Object {
  "dimensions": false,
  "expandProps": "end",
  "icon": false,
  "memo": false,
  "namedExport": "ReactComponent",
  "native": false,
  "plugins": null,
  "prettier": true,
  "prettierConfig": null,
  "ref": false,
  "replaceAttrValues": null,
  "runtimeConfig": true,
  "svgProps": null,
  "svgo": true,
  "svgoConfig": Object {
    "plugins": Array [
      Object {
        "active": true,
        "name": "convertStyleToAttrs",
      },
    ],
  },
  "template": null,
  "titleProp": false,
  "typescript": false,
}
`;

exports[`svgo sync #loadConfig [sync] should work with custom config path 1`] = `
Object {
  "dimensions": true,
  "expandProps": "end",
  "icon": true,
  "memo": false,
  "namedExport": "ReactComponent",
  "native": false,
  "noSemi": true,
  "plugins": null,
  "prettier": true,
  "prettierConfig": null,
  "ref": false,
  "replaceAttrValues": Array [
    Array [
      "#063855",
      "currentColor",
    ],
  ],
  "runtimeConfig": true,
  "svgProps": null,
  "svgo": true,
  "svgoConfig": Object {
    "plugins": Array [
      Object {
        "active": true,
        "name": "convertStyleToAttrs",
      },
    ],
  },
  "template": null,
  "titleProp": false,
  "typescript": false,
}
`;
